File: JavaScript/979-Distribute Coins in Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var distributeCoins = function(root) {
    let res = 0;
    let dfs = (node) =>{
        if (!node){
            return 0;
        }
        let left = dfs(node.left);
        let right = dfs(node.right);
        res += Math.abs(left) + Math.abs(right);
        return node.val - 1 + left + right;
    }
    dfs(root);
    return res;
};

File: C++/979-Distribute Coins in Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    int dfs(TreeNode* node) {
        if (node == nullptr) {
            return 0;
        }
        int left = dfs(node->left);
        int right = dfs(node->right);
        res += abs(left) + abs(right);

        return left + right + node->val - 1;  // how many coins we need (negative) or move out (positive) to make all nodes in the subtree have one coin
    }
    int distributeCoins(TreeNode* root) {
        dfs(root);
        return res;
    }
};